### ISSUE TITLE
exception type std::runtime_error: FIXME: bailing out here, need to fix this further up the call stack

### ISSUE THREAD

**Issue by PeeJay** [Zygo/bees#60]
I am getting quite a lot of the below exceptions - are they something I should worry about? I've screwed up this filesystem twice now trying to dedupe it, so I'm trying to find out what the trouble is. I've ruled out hardware as it was perfect until I tried deduping, and the same disks contain another filesystem which is operating correctly. (Two raid1s sharing the same pair of discs, 1/2 each.)
Kernel last time was 4.15.latest, this time is 4.14.24.

```
<6>crawl_258[23593]: scan: 4K 0x4aa4000 [D] 0x4aa5000 /media/backup/jones/2018-03-09/pj-google/Gigatech/invoice template.bmp
<6>crawl_258[23593]: dedup: src 4K [0x4aa4000..0x4aa5000] {0xa643e000z3000} /media/backup/jones/2018-03-09/pj-google/Gigatech/invoice template.bmp
<6>crawl_258[23593]:        dst 4K [0x4aa7000..0x4aa8000] {0xa6439000z7000} /media/backup/jones/2018-03-09/pj-google/Gigatech/invoice template.bmp
<6>crawl_258[23593]: dedup: src 4K [0x4a8a000..0x4a8b000] {0xa6421000z4000} /media/backup/jones/2018-03-09/pj-google/Gigatech/invoice template.bmp
<6>crawl_258[23593]:        dst 4K [0x4aa8000..0x4aa9000] {0xa6439000z8000} /media/backup/jones/2018-03-09/pj-google/Gigatech/invoice template.bmp
<3>crawl_258[23598]: Opening src bfr 4K [0x4806000..0x4807000] fid = 258:697 fd = 370 '/media/backup/jones/2018-03-09/pj-google/Gigatech/invoice template.bmp'
<3>crawl_258[23598]: chase_extent_ref ino BtrfsInodeOffsetRoot { .m_inum = 697, .m_offset = 0x4805000, .m_root = 258 } bbd BeesBlockData { 4K 0x4a43000 fd = 370 '/media/backup/jones/2018-03-09/pj-google/Gigatech/invoice template.bmp', data[4096] }
<3>crawl_258[23598]: ino_off_root BtrfsInodeOffsetRoot { .m_inum = 697, .m_offset = 0x4805000, .m_root = 258 }
<3>crawl_258[23598]: for_each_extent_ref BeesBlockData { 4K 0x4a43000 fd = 370 '/media/backup/jones/2018-03-09/pj-google/Gigatech/invoice template.bmp', data[4096] } at 0xa6421000z5000: 2053 found
<3>crawl_258[23598]: overlap_bfr 0 (..0x0]
<3>crawl_258[23598]: Opening dst bfr 4K [0x4a43000..0x4a44000] fd = 370 '/media/backup/jones/2018-03-09/pj-google/Gigatech/invoice template.bmp'
<3>crawl_258[23598]: replace_dst dst_bfr 4K [0x4a43000..0x4a44000] fd = 370 '/media/backup/jones/2018-03-09/pj-google/Gigatech/invoice template.bmp'
<3>crawl_258[23598]: finding one match (out of 2053) at 0xa6421000z5000 for BeesBlockData { 4K 0x4a43000 fd = 370 '/media/backup/jones/2018-03-09/pj-google/Gigatech/invoice template.bmp', address = 0xa6420000z7000, hash = 0x661fd10b39c0f81e, data[4096] }
<3>crawl_258[23598]: resolved_addrs.size() = 3
<3>crawl_258[23598]: Pushing hash 0x661fd10b39c0f81e addr 0xa6420000z7000 bbd BeesBlockData { 4K 0x4a43000 fd = 370 '/media/backup/jones/2018-03-09/pj-google/Gigatech/invoice template.bmp', address = 0xa6420000z7000, hash = 0x661fd10b39c0f81e, data[4096] }
<3>crawl_258[23598]: scan bbd BeesBlockData { 4K 0x4a43000 fd = 370 '/media/backup/jones/2018-03-09/pj-google/Gigatech/invoice template.bmp', address = 0xa6420000z7000, hash = 0x661fd10b39c0f81e, data[4096] }
<3>crawl_258[23598]: Extent { begin = 0x4a43000, end = 0x4a44000, physical = 0xa6420000, flags = Extent::OBSCURED|FIEMAP_EXTENT_ENCODED, physical_len = 0x20000, logical_len = 0x1000, offset = 0x7000 } block_count 1
<3>crawl_258[23598]: scan extent Extent { begin = 0x4a43000, end = 0x4a44000, physical = 0xa6420000, flags = Extent::OBSCURED|FIEMAP_EXTENT_ENCODED, physical_len = 0x20000, logical_len = 0x1000, offset = 0x7000 }
<3>crawl_258[23598]:
<3>crawl_258[23598]:
<3>crawl_258[23598]: *** EXCEPTION ***
<3>crawl_258[23598]:    exception type std::runtime_error: FIXME: bailing out here, need to fix this further up the call stack
<3>crawl_258[23598]: ***
<6>crawl_258[23595]: dedup: src 4K [0x4a67000..0x4a68000] {0xa6421000z2000} /media/backup/jones/2018-03-09/pj-google/Gigatech/invoice template.bmp
<6>crawl_258[23595]:        dst 4K [0x4aa0000..0x4aa1000] {0xa6420000z7000} /media/backup/jones/2018-03-09/pj-google/Gigatech/invoice template.bmp
<3>crawl_258[23593]: Opening src bfr 4K [0x47e7000..0x47e8000] fid = 258:697 fd = 370 '/media/backup/jones/2018-03-09/pj-google/Gigatech/invoice template.bmp'
<3>crawl_258[23593]: chase_extent_ref ino BtrfsInodeOffsetRoot { .m_inum = 697, .m_offset = 0x47e2000, .m_root = 258 } bbd BeesBlockData { 4K 0x4aa8000 fd = 370 '/media/backup/jones/2018-03-09/pj-google/Gigatech/invoice template.bmp', data[4096] }
<3>crawl_258[23593]: ino_off_root BtrfsInodeOffsetRoot { .m_inum = 697, .m_offset = 0x47e2000, .m_root = 258 }
<3>crawl_258[23593]: for_each_extent_ref BeesBlockData { 4K 0x4aa8000 fd = 370 '/media/backup/jones/2018-03-09/pj-google/Gigatech/invoice template.bmp', data[4096] } at 0xa6421000z7000: 2111 found
<3>crawl_258[23593]: overlap_bfr 0 (..0x0]
<3>crawl_258[23593]: Opening dst bfr 4K [0x4aa8000..0x4aa9000] fd = 370 '/media/backup/jones/2018-03-09/pj-google/Gigatech/invoice template.bmp'
<3>crawl_258[23593]: replace_dst dst_bfr 4K [0x4aa8000..0x4aa9000] fd = 370 '/media/backup/jones/2018-03-09/pj-google/Gigatech/invoice template.bmp'
<3>crawl_258[23593]: finding one match (out of 2111) at 0xa6421000z7000 for BeesBlockData { 4K 0x4aa8000 fd = 370 '/media/backup/jones/2018-03-09/pj-google/Gigatech/invoice template.bmp', address = 0xa6439000z8000, hash = 0xf8ad5e859508bb0d, data[4096] }
<3>crawl_258[23593]: resolved_addrs.size() = 2
<3>crawl_258[23593]: Pushing hash 0xf8ad5e859508bb0d addr 0xa6439000z8000 bbd BeesBlockData { 4K 0x4aa8000 fd = 370 '/media/backup/jones/2018-03-09/pj-google/Gigatech/invoice template.bmp', address = 0xa6439000z8000, hash = 0xf8ad5e859508bb0d, data[4096] }
<3>crawl_258[23593]: scan bbd BeesBlockData { 4K 0x4aa8000 fd = 370 '/media/backup/jones/2018-03-09/pj-google/Gigatech/invoice template.bmp', address = 0xa6439000z8000, hash = 0xf8ad5e859508bb0d, data[4096] }
<3>crawl_258[23593]: Extent { begin = 0x4aa5000, end = 0x4ac0000, physical = 0xa6439000, flags = Extent::OBSCURED|FIEMAP_EXTENT_ENCODED, physical_len = 0x20000, logical_len = 0x1b000, offset = 0x5000 } block_count 27
<3>crawl_258[23593]: scan extent Extent { begin = 0x4aa5000, end = 0x4ac0000, physical = 0xa6439000, flags = Extent::OBSCURED|FIEMAP_EXTENT_ENCODED, physical_len = 0x20000, logical_len = 0x1b000, offset = 0x5000 }
<3>crawl_258[23593]:
<3>crawl_258[23593]:
<3>crawl_258[23593]: *** EXCEPTION ***
<3>crawl_258[23593]:    exception type std::runtime_error: FIXME: bailing out here, need to fix this further up the call stack
<3>crawl_258[23593]: ***
<3>crawl_258[23595]: Opening src bfr 4K [0x4806000..0x4807000] fid = 258:697 fd = 370 '/media/backup/jones/2018-03-09/pj-google/Gigatech/invoice template.bmp'
<3>crawl_258[23595]: chase_extent_ref ino BtrfsInodeOffsetRoot { .m_inum = 697, .m_offset = 0x4805000, .m_root = 258 } bbd BeesBlockData { 4K 0x4aa0000 fd = 370 '/media/backup/jones/2018-03-09/pj-google/Gigatech/invoice template.bmp', data[4096] }
<3>crawl_258[23595]: ino_off_root BtrfsInodeOffsetRoot { .m_inum = 697, .m_offset = 0x4805000, .m_root = 258 }
<3>crawl_258[23595]: for_each_extent_ref BeesBlockData { 4K 0x4aa0000 fd = 370 '/media/backup/jones/2018-03-09/pj-google/Gigatech/invoice template.bmp', data[4096] } at 0xa6421000z5000: 2174 found
<3>crawl_258[23595]: overlap_bfr 0 (..0x0]
<3>crawl_258[23595]: Opening dst bfr 4K [0x4aa0000..0x4aa1000] fd = 370 '/media/backup/jones/2018-03-09/pj-google/Gigatech/invoice template.bmp'
<3>crawl_258[23595]: replace_dst dst_bfr 4K [0x4aa0000..0x4aa1000] fd = 370 '/media/backup/jones/2018-03-09/pj-google/Gigatech/invoice template.bmp'
<3>crawl_258[23595]: finding one match (out of 2174) at 0xa6421000z5000 for BeesBlockData { 4K 0x4aa0000 fd = 370 '/media/backup/jones/2018-03-09/pj-google/Gigatech/invoice template.bmp', address = 0xa6420000z7000, hash = 0x661fd10b39c0f81e, data[4096] }
<3>crawl_258[23595]: resolved_addrs.size() = 3
<3>crawl_258[23595]: Pushing hash 0x661fd10b39c0f81e addr 0xa6420000z7000 bbd BeesBlockData { 4K 0x4aa0000 fd = 370 '/media/backup/jones/2018-03-09/pj-google/Gigatech/invoice template.bmp', address = 0xa6420000z7000, hash = 0x661fd10b39c0f81e, data[4096] }
<3>crawl_258[23595]: scan bbd BeesBlockData { 4K 0x4aa0000 fd = 370 '/media/backup/jones/2018-03-09/pj-google/Gigatech/invoice template.bmp', address = 0xa6420000z7000, hash = 0x661fd10b39c0f81e, data[4096] }
<3>crawl_258[23595]: Extent { begin = 0x4aa0000, end = 0x4aa1000, physical = 0xa6420000, flags = Extent::OBSCURED|FIEMAP_EXTENT_ENCODED, physical_len = 0x20000, logical_len = 0x1000, offset = 0x7000 } block_count 1
<3>crawl_258[23595]: scan extent Extent { begin = 0x4aa0000, end = 0x4aa1000, physical = 0xa6420000, flags = Extent::OBSCURED|FIEMAP_EXTENT_ENCODED, physical_len = 0x20000, logical_len = 0x1000, offset = 0x7000 }
<3>crawl_258[23595]:
<3>crawl_258[23595]:
<3>crawl_258[23595]: *** EXCEPTION ***
<3>crawl_258[23595]:    exception type std::runtime_error: FIXME: bailing out here, need to fix this further up the call stack
<3>crawl_258[23595]: ***
<6>crawl_258[23599]: dedup: src 4K [0x4aa0000..0x4aa1000] {0xa6421000z2000} /media/backup/jones/2018-03-09/pj-google/Gigatech/invoice template.bmp
<6>crawl_258[23599]:        dst 4K [0x4a82000..0x4a83000] {0xa6420000z7000} /media/backup/jones/2018-03-09/pj-google/Gigatech/invoice template.bmp

```

**Comment by Zygo** [comment]
The current bees algorithm loops too many times when many blocks in many files are identical.  This is fairly common in BMP files.  The exception just breaks the loop.  If the loop is allowed to run until completion it can take several days to process each file.

Can you say more about what exactly is going wrong with the filesystem?

**Comment by PeeJay** [comment]
When it first happened I asked on the BTRFS mailing list for help to recover (https://www.spinics.net/lists/linux-btrfs/msg75143.html) but I was using zstd compression so couldn't use any of the normal tools, so I gave up. That instance of corruption was particularly bad, as it was only noticed after the FS went read only.
The second time (during the rebuild) I was being a bit more proactive about running scrub every day or two, which is how I found the corruption. There were no IO errors or other kernel messages beforehand. I wasn't capturing the output from bees so I don't know if it emitted any errors.
I tried to fix the FS, but btrfs check couldn't do it. Then I tried to btrfs check --init-extent-tree which seemed to be working, but I stopped it after millions of errors and 26 hours.
I'm rebuilding for the 3rd time now and no errors so far. I haven't deleted any snapshots with bees running (I remember a while back that would crash bees). Just create snapshot, rsync data, and repeat.
I may have also rebooted while bees was running, and if the drive didn't unmount in time then a bunch of dirty metadata would get lost.
Basically I have no idea where the problem is at the moment, still looking. I doubt it's bees though, more likely the kernel dedupe ioctls. If I have anything to report I'll let you know
